File: src/advent/day_11/part_2.hs
module Part2 where

import Data.Function (fix)
import System.IO (readFile)

getNumberOfDigits :: Int -> Int
getNumberOfDigits number = length $ show number

getNumberOfStones :: (Int -> Int -> Int) -> Int -> Int -> Int
getNumberOfStones f _ 0 = 1
getNumberOfStones f number iterations
  | number == 0 = f 1 (iterations - 1)
  | even $ getNumberOfDigits number = let (firstHalf, secondHalf) = splitAt (length (show number) `div` 2) (show number)
                                       in f (read firstHalf) (iterations - 1) + f (read secondHalf) (iterations - 1)
  | otherwise = f (number * 2024) (iterations - 1)

memoize :: (Int -> a) -> (Int -> a)
memoize f = (map f [0 ..] !!)

getNumberOfStonesMemoized :: Int -> Int -> Int
getNumberOfStonesMemoized = fix (memoize . getNumberOfStones)

solution :: FilePath -> IO Int
solution filePath = do
  contents <- readFile filePath
  let numbers = map read $ words contents
  return $ sum $ map (`getNumberOfStonesMemoized` 75) numbers

main :: IO ()
main = do
    result <- solution "data.txt"
    print result

File: src/advent/day_11/part_2.py
from functools import cache
from pathlib import Path

from utils.load_file import File

def get_number_of_didigts(number: int) -> int:
    return len(str(number))

@cache
def get_number_of_stones(number: int, iterations: int) -> int:
    if iterations > 0:
        if number == 0:
            return get_number_of_stones(1, iterations - 1)
        elif get_number_of_didigts(number) % 2 == 0:
            first_half, second_half = str(number)[:len(str(number)) // 2], str(number)[len(str(number)) // 2:]
            return get_number_of_stones(int(first_half), iterations - 1) + get_number_of_stones(int(second_half), iterations - 1)
        else:
            return get_number_of_stones(number * 2024, iterations - 1)
    else:
        return 1

def solution(file_path: Path) -> int:
    numbers: list[int] = []
    
    for line in File(file_path).read():
        line = line.strip()
        if line == "":
            break

        numbers = list(map(int, line.split()))

    return sum(get_number_of_stones(number, 75) for number in numbers)

def main() -> None:
    assert solution("day_11/data_test.txt") == 65601038650482
    print(solution("day_11/data.txt"))

if __name__ == "__main__":
    main()